### PR TITLE
Add cluster-api integration enhancement

### PR DESCRIPTION
**Pull request by alexander-demichev**

### CODE REVIEW

File: enhancements/machine-api/cluster-api-integration.md [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement describes the process of integrating the upstream [Cluster API](https://github.com/kubernetes-sigs/cluster-api)
+project into OpenShift standalone clusters.
+
+## Motivation
+
+We would like to give users the ablility to use Cluster API for machine management, as an alternative and possible future replacement for Machine API.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:37]: should we perhaps say that we want to give users the ability to use CAPI as an _addition or supplement_ for machine management?

i don't think we are actually giving the ability to use CAPI as an alternative to MAPI at this point, or will a user be able to replace their initial MAPI MachineSets with CAPI MachineSets?

for example,

i am an OpenShift user, i do a normal install (now i have 3 Machines and 3 MachineSets managed by MAPI). i enable the CAPI tech preview and create 3 MachineSets each with 1 replica. can i now delete my original 3 worker nodes that are managed by MAPI?

File: enhancements/machine-api/cluster-api-integration.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+- Ensure feature parity between MAPI and CAPI before migration
+
+### Non-Goals
+
+- Deprecate or remove any existing APIs
+- Stop providing support for Machine API in near future
+
+## Proposal
+
+In the longer term, our goal is to reconverge with the upstream Cluster API project and leverage the community built providers for managing Machines within OpenShift standalone clusters.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:52]: i'm a little hesitant about stating our goal to reconverge with CAPI here. afaik we are still evaluating making that choice. especially given that the following sentences talk about using this as an evaluation.

File: enhancements/machine-api/cluster-api-integration.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+To gather feedback on the feasability of this goal, we will first introduce Cluster API alongside Machine API as a technical preview, and gather feedback on Cluster APIs feasability as a
+replacement for Machine API.
+
+### User Stories
+
+#### Story 1
+
+As an openshift developer, I would like to leverage the upstream providers and reduce the barrier to OpenShift of supporting new platforms.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:60]: minor clarification suggestion
```suggestion
As an OpenShift developer, I would like to leverage the upstream community Cluster API infrastructure providers and reduce the barrier to OpenShift of supporting new platforms.
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+#### Story 2
+
+As an openshift developer, I would like to gain the support of third parties who already have vested interests in maintaining Machine providers for various platforms and environments and benefit from their provider specific expertise as I try to add new features.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:64]: maybe say we would like to collaborate with more third parties rather than gain support?

File: enhancements/machine-api/cluster-api-integration.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+#### Story 3
+
+As a cloud developer, I would like to easily onboard new platforms as this process is well documented and any implementation will be able to be leveraged by both Kubernetes and OpenShift customers, increasing the value of implementing a new provider.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:68]: i might mention the CAPI community in here just to make it clear that we are using that community's documentation and guidance to create those new infrastructure providers.

File: enhancements/machine-api/cluster-api-integration.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+As a developer, I would like to be able to use the same set of tools for infrastructure management in OpenShift as I can for vanilla Kubernetes.
+
+### Implementation Details
+
+First, we need to establish Cluster API resource management by ensuring all required components are successfully installed and running within the OpenShift cluster.
+
+Cluster API will only be present in the cluster (installed by a new operator) if and when a user installs a feature gate.
+We will introduce a new, OpenShift specific, feature gate `ClusterAPIEnabled` and include it within the `TechPreviewNoUpgrade` FeatureSet.
+
+Once installed, this preview will allow the user to create new MachineSets using CAPI and explore the features available within CAPI, for comparison with MAPI.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:81]: will MachineDeployments be available as well?

**Review by elmiko — COMMENTED** [review]
good start, i have a few questions inline, also i'm curious about how a user might use CAPI on OpenShift. will we allow users to create new clusters from the CAPI on their OpenShift install? (essentially allowing them to use OpenShift as the management cluster for new workload clusters)

File: enhancements/machine-api/cluster-api-integration.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
+---
+title: Cluster API Integration
+authors:
+  - "@JoelSpeed"
+  - "@alexander-demichev"
+reviewers:
+  - "@elmiko"
+  - "@Fedosin"
+  - "@lobziik"
+  - "asalkeld"
```
Comment (asalkeld) [inline enhancements/machine-api/cluster-api-integration.md:10]: @asalkeld

File: enhancements/machine-api/cluster-api-integration.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+  - "hardys"
```
Comment (asalkeld) [inline enhancements/machine-api/cluster-api-integration.md:11]:
```suggestion
  - "@hardys"
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+- Deprecate or remove any existing APIs
+- Stop providing support for Machine API in near future
+
+## Proposal
+
+In the longer term, our goal is to reconverge with the upstream Cluster API project and leverage the community built providers for managing Machines within OpenShift standalone clusters.
+To gather feedback on the feasability of this goal, we will first introduce Cluster API alongside Machine API as a technical preview, and gather feedback on Cluster APIs feasability as a
+replacement for Machine API.
+
+### User Stories
```
Comment (asalkeld) [inline enhancements/machine-api/cluster-api-integration.md:56]: Is it worth stating the major functional benefits of CAPI? As these are gaps that CAPI has over MAPI..
- support for mixed platform clusters
- support for hub-spoke clusters

Comment (hardys) [reply enhancements/machine-api/cluster-api-integration.md:56]: IMO it is worth calling these out - the other user stories are all "As a developer ..." - which is an appropriate persona for the initial preview phase of work, but these two features are high priority user-facing benefits in the long term.

Maybe we could describe those in "As a user ..." terms, but then make clear that this is just a first step towards enabling those use-cases, not the final solution?

File: enhancements/machine-api/cluster-api-integration.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+
+##### Deploy core Cluster API
+
+Once the upstream Cluster API Operator is installed, the next step is to create a `CoreProvider` CR along with a configmap that contains upstream Cluster API CRDs, Deployment, Webhooks and RBAC resources.
+The example usage is described [here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md#air-gapped-environment).
+
+##### Deploy Cluster API infrastructure provider
+
+The `cluster-capi-operator` operator will create an appropriate `InfrastructureProvider` CR (based on the cluster platform) and a configmap that contains upstream Cluster API cloud provider CRDs, Deployment, Webhooks and RBAC resources.
+
```
Comment (hardys) [inline enhancements/machine-api/cluster-api-integration.md:120]: Do we expect to target only specific platforms/providers in this initial preview, or all currently supported platforms?

File: enhancements/machine-api/cluster-api-integration.md (hunk 2 above)
Comment (JoelSpeed) [reply enhancements/machine-api/cluster-api-integration.md:52]: I think you're right, this is an evaluation tool for us, to see if we think this is feasible and assess feedback on running CAPI within OCP. Though I don't think there's necessarily anything wrong with this sentence. The reason we are running this TechPreview is to evaluate whether our longer term goal of reconverging is viable.

Is there a way you would reword this to make it more clear? The next sentence is talking about how this preview allows us to gather feedback on the feasability of the goal

File: enhancements/machine-api/cluster-api-integration.md (hunk 1 above)
Comment (flaper87) [reply enhancements/machine-api/cluster-api-integration.md:37]: I find it a bit confusing that the description of the enhancement mentions `possible future replacement for Machine API` but then `Stop providing support for Machine API in near future` is stated as a non-Goal. I did notice that `near` in the non-Goal, which helps understanding the intention.

For the sake of scope and clarity, I would rather state that the intention is to replace MAPI in the future (basically without the `possible`) or just don't mention it at all. I lean more towards not mentioning the replacement at all, as that also aligns with the scenario metioned by @elmiko  (:wave: )

File: enhancements/machine-api/cluster-api-integration.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+First, we need to establish Cluster API resource management by ensuring all required components are successfully installed and running within the OpenShift cluster.
+
+Cluster API will only be present in the cluster (installed by a new operator) if and when a user installs a feature gate.
+We will introduce a new, OpenShift specific, feature gate `ClusterAPIEnabled` and include it within the `TechPreviewNoUpgrade` FeatureSet.
+
+Once installed, this preview will allow the user to create new MachineSets using CAPI and explore the features available within CAPI, for comparison with MAPI.
+
+During this timeframe, any user wanting to migrate or try out the preview will be left to manually migrate the MachineSet, using either the upstream documentation or documentation provider by OpenShift.
+
+This preview will be intended to be used to create day 2 worker MachineSets and is not expected to be integrated into the install process in any way.
```
Comment (flaper87) [inline enhancements/machine-api/cluster-api-integration.md:85]: This almost sounds like a non-goal to me:

- Provide any automated integration or migration between MAPI and CAPI resources.

File: enhancements/machine-api/cluster-api-integration.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+#### Cluster API resource management
+
+In order to maintain the lifecycle of Cluster API related resources, we will create a new operator `cluster-capi-operator`, this name was chosen for avoiding confusion with upstream Cluster API operator.
```
Comment (flaper87) [inline enhancements/machine-api/cluster-api-integration.md:89]:
```suggestion
In order to maintain the lifecycle of Cluster API related resources, we will create a new operator `openshift-cluster-api-operator`, this name was chosen for avoiding confusion with upstream Cluster API operator.
```

Sorry for starting a bikeshed on the name, but I think in this case it's important.

`cluster-capi-operator` will still confuse users. The name won't clash with the upstream one but it will create confusion on which is responsible for what.

I'd propose making it clear that `cluster-capi-operator` is an OpenShift specific thing (maybe mention OCP in the name?). I don't know what the OpenShift naming conventions in this case are so I will let you pick and choose the best name.

File: enhancements/machine-api/cluster-api-integration.md [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+##### Reconcile FeatureGate object
+
+While Cluster API intergration is in tech preview, the operator will reconcile the cluster [`FeatureGate`](https://docs.openshift.com/container-platform/4.8/nodes/clusters/nodes-cluster-enabling-features.html) object and check for `ClusterAPIEnabled` feature gate presence.
+The operator will procceed with Cluster API installation if and only if the required feature gate is present.
+
+##### Deploy Cluster Machine Approver
+
+In order for Cluster API machines to succefully join the cluster, the Kubelet CSRs need to be approved.
+The operator will deploy a separate instance of the cluster-machine-approver, which will be configured to be used with Cluster API machines by providing [`--apigroup`](https://github.com/openshift/cluster-machine-approver/blob/master/main.go#L54) flag that was recently introduced.
```
Comment (flaper87) [inline enhancements/machine-api/cluster-api-integration.md:105]: Will this be temporary (while the feature is gated)? Or is it expected that when using CAPI with OCP there will always be 2 machine approvers?

File: enhancements/machine-api/cluster-api-integration.md (hunk 10 above)
Comment (flaper87) [reply enhancements/machine-api/cluster-api-integration.md:120]: Probably worth limiting to a few? What would the test matrix for this look like if all the currently supported platforms are enabled?

File: enhancements/machine-api/cluster-api-integration.md [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+      - filters:
+        - name: tag:Name
+          values:
+          - ...
+```
+
+### Risks and Mitigations
+
+- During tech preview `cluster-capi-operator` will have permissions to manage CRDs, this might be a not secure permission for an operator.
+- Note, this permission should be restricted to creating CRDs only, as once installed, the technical preview cannot be uninstalled.
```
Comment (flaper87) [inline enhancements/machine-api/cluster-api-integration.md:217]: I'm assuming this is needed so that the CAPI CRDs can be installed. If so, can we list the permissions and motivations here explicitly? In my experience, this helps drive the conversation and with future reviews.

Comment (asalkeld) [reply enhancements/machine-api/cluster-api-integration.md:217]: currently the upstream operator has clusterRoles of "*" for api/resources/verbs as it installs roles itself and you can't install
a role that you don't have yourself, because of this cluster-capi-operator also has this (in my PR https://github.com/openshift/cluster-capi-operator/pull/8)

File: enhancements/machine-api/cluster-api-integration.md (hunk 1 above)
Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:37]: You are right, I will not mention MAPI replacement in this proposal. I think the process of MAPI migration to CAPI might require a separate proposal.

File: enhancements/machine-api/cluster-api-integration.md (hunk 10 above)
Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:120]: I will add a section about supported platforms that will definitely be in tech preview, but some other platforms might be added later.

File: enhancements/machine-api/cluster-api-integration.md (hunk 11 above)
Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:85]: I will add this as a non-goal, at this stage we expect users to have CAPI resources running together with MAPI.

File: enhancements/machine-api/cluster-api-integration.md (hunk 12 above)
Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:89]: `cluster-capi-operator` is the less confusing name we came up with. The repo is located in openshift org https://github.com/openshift/cluster-capi-operator, this makes it clear that the operator belongs to openshift.

File: enhancements/machine-api/cluster-api-integration.md [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+### Implementation Details
+
+First, we need to establish Cluster API resource management by ensuring all required components are successfully installed and running within the OpenShift cluster.
+
+Cluster API will only be present in the cluster (installed by a new operator) if and when a user installs a feature gate.
+We will introduce a new, OpenShift specific, feature gate `ClusterAPIEnabled` and include it within the `TechPreviewNoUpgrade` FeatureSet.
+
+Once installed, this preview will allow the user to create new MachineSets using CAPI and explore the features available within CAPI, for comparison with MAPI.
+
+During this timeframe, any user wanting to migrate or try out the preview will be left to manually migrate the MachineSet, using either the upstream documentation or documentation provider by OpenShift.
```
Comment (enxebre) [inline enhancements/machine-api/cluster-api-integration.md:91]: typo "provider"/"provided"

File: enhancements/machine-api/cluster-api-integration.md [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+This preview will be intended to be used to create day 2 worker MachineSets and is not expected to be integrated into the install process in any way.
+
+### Supported platforms
+
+The techincal preview aims to support: AWS, Azure, GCP.
+
+#### Cluster API resource management
+
+In order to maintain the lifecycle of Cluster API related resources, we will create a new operator `cluster-capi-operator`, this name was chosen for avoiding confusion with upstream Cluster API operator.
+This operator will be responsible for all administrative tasks related to the deployment of the Cluster API project within the cluster.
```
Comment (enxebre) [inline enhancements/machine-api/cluster-api-integration.md:102]: >This operator will be responsible for all administrative tasks related to the deployment of the Cluster API project within the cluster.

can you enumerate and elaborate these tasks?

Comment (enxebre) [reply enhancements/machine-api/cluster-api-integration.md:102]: nvm I see them below

File: enhancements/machine-api/cluster-api-integration.md [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+
+Once installed, this preview will allow the user to create new MachineSets using CAPI and explore the features available within CAPI, for comparison with MAPI.
+
+During this timeframe, any user wanting to migrate or try out the preview will be left to manually migrate the MachineSet, using either the upstream documentation or documentation provider by OpenShift.
+
+This preview will be intended to be used to create day 2 worker MachineSets and is not expected to be integrated into the install process in any way.
+
+### Supported platforms
+
+The techincal preview aims to support: AWS, Azure, GCP.
```
Comment (asalkeld) [inline enhancements/machine-api/cluster-api-integration.md:97]: not metal3? We have customers that are very keen have virt controlplane and metal3 workers

File: enhancements/machine-api/cluster-api-integration.md [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+`cluster-capi-operator` should first install the [upstream operator CRDs](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md#new-api-types) and then run the upstream operator itself using a `Deployment`.
+
+##### Deploy core Cluster API
+
+Once the upstream Cluster API Operator is installed, the next step is to create a `CoreProvider` CR along with a configmap that contains upstream Cluster API CRDs, Deployment, Webhooks and RBAC resources.
+The example usage is described [here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md#air-gapped-environment).
+
+##### Deploy Cluster API infrastructure provider
+
+The `cluster-capi-operator` operator will create an appropriate `InfrastructureProvider` CR (based on the cluster platform) and a configmap that contains upstream Cluster API cloud provider CRDs, Deployment, Webhooks and RBAC resources.
```
Comment (asalkeld) [inline enhancements/machine-api/cluster-api-integration.md:131]: one question I have is many of the providers require a secret for the variables used in the manifests.
Do we need platform specific code to gather this and create the secret so the the infrastructure provider can
start correctly? Currently I am generating defaults only and creating those secrets, but the user would have to edit the secret if the defaults are not good enough.

File: enhancements/machine-api/cluster-api-integration.md (hunk 17 above)
Comment (sadasu) [reply enhancements/machine-api/cluster-api-integration.md:97]: Yes, we would like Baremetal to be one of the supported platfoms early on in the tech preview. Do you see any challenges to that?

File: enhancements/machine-api/cluster-api-integration.md (hunk 15 above)
Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:91]: they can, I will add it.

File: enhancements/machine-api/cluster-api-integration.md [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+A new `cluster-capi-operator` image will be built and included in every release payload.
+
+#### Credentials management
+
+The `cluster-capi-operator`'s manifests should contain an appropriate `CredentialsRequest` for each supported cloud provider.
+This is similiar to [machine-api-operator](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml)
+
+#### Cluster API cloud providers
+
+Cluster API cloud providers will live in a forks, similar to what is now done for Machine API.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:166]:
```suggestion
Cluster API cloud providers will live in forks, similar to what is now done for Machine API.
```

also, i think we should mention something about how we will name these forks and if we will want to change our current cluster-api-* repos.

File: enhancements/machine-api/cluster-api-integration.md [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+        - name: tag:Name
+          values:
+          - ...
+```
+
+### Risks and Mitigations
+
+- During tech preview `cluster-capi-operator` will have permissions to manage CRDs, this might be a not secure permission for an operator.
+- Note, this permission should be restricted to creating CRDs only, as once installed, the technical preview cannot be uninstalled.
+- CLI usage, once Cluster API is installed command like `oc get machine` will return Cluster API machines, in order to use Machine API users will have to use fully qualified name `oc get machines.machine.openshift.io`.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:231]: this makes me wonder if we shouldn't have some sort of warning message associated with `oc get machines` once the feature gate is active, but i'm not sure if that's even possible.

**Review by elmiko — COMMENTED** [review]
i think this is looking better @alexander-demichev , i left a couple comments. also, it might be helpful to start resolving some of the conversations to make the reviewing a little easier.

File: enhancements/machine-api/cluster-api-integration.md (hunk 20 above)
Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:231]: Yeah, we still have to figure out what to do here.

File: enhancements/machine-api/cluster-api-integration.md [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+The `cluster-capi-operator` will need to create the `Cluster` and a proper `InfrastructureCluster` resource for the OpenShift cluster.
+Because we have our own infrastructure management strategy in OpenShift, we should leverage the [externally managed cluster infrastructure](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20210203-externally-managed-cluster-infrastructure.md) feature.
+This means that the created `InfrastructureCluster` should have `cluster.x-k8s.io/managed-by:` annotation set.
+
+Cluster object reconciler can be done as a separate controller. The controller should:
+- Wait before Cluster and InfrastructureCluster CRD is present
+- Create both Cluster and InfrastructureCluster objects with externally managed cluster infrastructure annotation.
+- Ensure spec/status of InfrastructureCluster are configured for the OpenShift cluster (infrastructure information can be sourced from resources within the OpenShift Cluster).
+- Patch `Cluster` status to `Ready=true`.
+
```
Comment (enxebre) [inline enhancements/machine-api/cluster-api-integration.md:158]: I don't see any mention to the control plane CR here? Which implementation are you planning to use?

File: enhancements/machine-api/cluster-api-integration.md [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+The `cluster-capi-operator` will need to create the `Cluster` and a proper `InfrastructureCluster` resource for the OpenShift cluster.
+Because we have our own infrastructure management strategy in OpenShift, we should leverage the [externally managed cluster infrastructure](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20210203-externally-managed-cluster-infrastructure.md) feature.
+This means that the created `InfrastructureCluster` should have `cluster.x-k8s.io/managed-by:` annotation set.
+
+Cluster object reconciler can be done as a separate controller. The controller should:
+- Wait before Cluster and InfrastructureCluster CRD is present
+- Create both Cluster and InfrastructureCluster objects with externally managed cluster infrastructure annotation.
+- Ensure spec/status of InfrastructureCluster are configured for the OpenShift cluster (infrastructure information can be sourced from resources within the OpenShift Cluster).
+- Patch `Cluster` status to `Ready=true`.
```
Comment (enxebre) [inline enhancements/machine-api/cluster-api-integration.md:157]: Who is creating the namespace where the cluster and machinesets CRs are mean to live?
This needs to be handled in a way that does not class with hypershift CAPI CRDs and controllers

**Comment by enxebre** [comment]
please add me as a reviewer/approver :)

File: enhancements/machine-api/cluster-api-integration.md (hunk 22 above)
Comment (asalkeld) [reply enhancements/machine-api/cluster-api-integration.md:157]: @enxebre does hypershift re-implement the capi controllers or does it install the capi controllers?
What I am getting at is, should they be able to work together (ignoring version differences), or can you deal with capi moving to openshift-cluster-api?

The CRDs have webhooks specifying a namespace (like this https://github.com/kubernetes-sigs/cluster-api/blob/main/config/crd/patches/webhook_in_clusters.yaml#L17) that will be different in the CRDs and WebHookConfigurations

Currently the downstream operator gathers and specifies the resources (including CRDs) and the upstream operator will install them.

File: enhancements/machine-api/cluster-api-integration.md (hunk 21 above)
Comment (asalkeld) [reply enhancements/machine-api/cluster-api-integration.md:158]: we are currently not installing the controlplane CR (but super easy to do so).

Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:158]: We are not planning to support control planes for tech preview

File: enhancements/machine-api/cluster-api-integration.md (hunk 22 above)
Comment (enxebre) [reply enhancements/machine-api/cluster-api-integration.md:158]: They should be able to operate independently. Hypershift owns the lifecycle of their CAPI controllers/CRs.

Comment (alexander-demichev) [reply enhancements/machine-api/cluster-api-integration.md:158]: We will only reconcile CAPI CRs in `openshift-cluster-api` namespace, CRDs can possibly conflict if we use different CAPI versions.

File: enhancements/machine-api/cluster-api-integration.md [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+
+### Supported platforms
+
+The technical preview aims to support: AWS, Azure, GCP, Baremetal, Openstack.
+
+#### Cluster API resource management
+
+In order to maintain the lifecycle of Cluster API related resources, we will create a new operator `cluster-capi-operator`, this name was chosen for avoiding confusion with upstream Cluster API operator.
+This operator will be responsible for all administrative tasks related to the deployment of the Cluster API project within the cluster.
+During tech preview phase, the new operator will also manage all Cluster API related CRDs. All CRD manifests will placed in
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:116]: minor nit
```suggestion
During tech preview phase, the new operator will also manage all Cluster API related CRDs. All CRD manifests will be placed in
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 24]
```diff
@@ -0,0 +1,305 @@
...
+openshift forks of CAPI and will take from there with no changes.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:117]: minor nit
```suggestion
OpenShift forks of CAPI and will take from there with no changes.
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 25]
```diff
@@ -0,0 +1,305 @@
...
+
+##### Reconcile FeatureGate object
+
+While Cluster API intergration is in tech preview, the operator will reconcile the cluster [`FeatureGate`](https://docs.openshift.com/container-platform/4.8/nodes/clusters/nodes-cluster-enabling-features.html) object and check for `ClusterAPIEnabled` feature gate presence.
+The operator will procceed with Cluster API installation if and only if the required feature gate is present.
+
+##### Deploy Cluster Machine Approver
+
+In order for Cluster API machines to succefully join the cluster, the Kubelet CSRs need to be approved.
+The operator will deploy a separate instance of the cluster-machine-approver, which will be configured to be used with Cluster API machines by providing [`--apigroup`](https://github.com/openshift/cluster-machine-approver/blob/master/main.go#L54) flag that was recently introduced.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:131]: since we use `code blocks` around the capi operator references earlier, i think we should use them here for the cma
```suggestion
The operator will deploy a separate instance of the `cluster-machine-approver`, which will be configured to be used with Cluster API machines by providing [`--apigroup`](https://github.com/openshift/cluster-machine-approver/blob/master/main.go#L54) flag that was recently introduced.
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 26]
```diff
@@ -0,0 +1,305 @@
...
+
+##### Install upstream CAPI operator
+
+We will use the upstream [Cluster API operator](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md) for managing CRDs and deploying cloud providers.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:135]: minor nit for consistency
```suggestion
We will use the upstream [Cluster API operator](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md) for managing CRDs and deploying infrastructure providers.
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 27]
```diff
@@ -0,0 +1,305 @@
...
+
+At this point all Cluster API components should be installed and ready to use.
+
+#### CVO management
+
+A new `cluster-capi-operator` image will be built and included in every release payload.
+
+#### Credentials management
+
+The `cluster-capi-operator`'s manifests should contain an appropriate `CredentialsRequest` for each supported cloud provider.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:174]: minor nit for consistency
```suggestion
The `cluster-capi-operator`'s manifests should contain an appropriate `CredentialsRequest` for each supported infrastructure provider.
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 28]
```diff
@@ -0,0 +1,305 @@
...
+This is similiar to [machine-api-operator](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml)
+
+#### Cluster API cloud providers
+
+Cluster API cloud providers will live in forks, similar to what is now done for Machine API. We now evaluating moving
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:179]: minor nit for consistency
```suggestion
Cluster API infrastructure providers will live in forks, similar to what is now done for Machine API. We now evaluating moving
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 29]
```diff
@@ -0,0 +1,305 @@
...
+- CLI usage, once Cluster API is installed command like `oc get machine` will return Cluster API machines, in order to use Machine API users will have to use fully qualified name `oc get machines.machine.openshift.io`.
+If we want to not introduce this breaking change then we have to set prefered API group in our [kubernetes fork](https://github.com/openshift/kubernetes/blob/master/pkg/controlplane/controller/crdregistration/patch.go).
+- Feature parity, for last year we've been trying to upstream all features introduced to Machine API but we can't be sure all of them work in upstream. We need to have a good set of regression tests running periodically.
+
+### API Extensions
+
+With `ClusterAPIEnabled` feature enabled, the following API extensions will be added:
+
+- Core Cluster API resources and webhooks, they can be found [here](https://github.com/kubernetes-sigs/cluster-api/tree/main/api/v1beta1)
+- Depending on a platform where cluster is running, infrastructure provider CRD and webhooks will be added, see
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:255]: minor nit for consistency
```suggestion
- Depending on the provider where a cluster is running, infrastructure provider CRD and webhooks will be added, see
```

File: enhancements/machine-api/cluster-api-integration.md [hunk 30]
```diff
@@ -0,0 +1,305 @@
...
+- Ability to have Cluster API installed using a feature gate
+- Ability to use Cluster API for machine management
+- End user documentation
+- Running upstream e2e workflow on openshift
+
+#### Tech Preview -> GA (Future Work)
+
+- Cluster API will be installed in all OpenShift clusters by default.
+- Bidirectional migration for MAPI and CAPI.
+- New cloud providers implemented as CAPI.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:291]: minor nit for consistency
```suggestion
- New infrastructure providers implemented as CAPI.
```

**Review by elmiko — COMMENTED** [review]
@alexander-demichev overall, i think the content here is good. i made a bunch of suggestions that i think will tighten up the language around the use of words like "cloud provider" and "platform". i also made some suggestions around spelling and grammar.

my requested changes here are cosmetic mostly, i think the core content is good.

File: enhancements/machine-api/cluster-api-integration.md [hunk 31]
```diff
@@ -0,0 +1,304 @@
...
+
+### Supported platforms
+
+The technical preview aims to support: AWS, Azure, GCP, Baremetal, Openstack.
+
+#### Cluster API resource management
+
+In order to maintain the lifecycle of Cluster API related resources, we will create a new operator `cluster-capi-operator`, this name was chosen for avoiding confusion with upstream Cluster API operator.
+This operator will be responsible for all administrative tasks related to the deployment of the Cluster API project within the cluster.
+During tech preview phase, the new operator leverage new [CVO feature](https://github.com/openshift/enhancements/blob/master/enhancements/update/cvo-techpreview-manifests.md) for managing all Cluster API related CRDs.
```
Comment (elmiko) [inline enhancements/machine-api/cluster-api-integration.md:117]: minor nit
```suggestion
During tech preview phase, the new operator will leverage the new [CVO feature](https://github.com/openshift/enhancements/blob/master/enhancements/update/cvo-techpreview-manifests.md) for managing all Cluster API related CRDs.
```

**Review by elmiko — COMMENTED** [review]
i think this generally reads well to me

the only question i have, which i don't think really fits within the enhancement scope is how the capi deployment might interact with the cluster-autoscaler. by default i don't imagine it would make any difference, but there is now the possibility for a user to install the autoscaler into a namespace that could read the capi machinesets and thus work in parallel with the openshift autoscaler.

just a though, i'm not sure we need to mention it here except to maybe say that we won't change the autoscaler behavior.

**Comment by alexander-demichev** [comment]
@elmiko I added a note about autoscaler to nongoals of this proposal.

File: enhancements/machine-api/cluster-api-integration.md (hunk 21 above)
Comment (elmiko) [reply enhancements/machine-api/cluster-api-integration.md:159]: might be nice to put this in the non-goals as well

**Review by elmiko — COMMENTED** [review]
generally looks good to me @alexander-demichev , i have a couple minor nits
/approve

**Comment by JoelSpeed** [comment]
This enhancement has been stagnant for a while with no additional feedback and no new discoveries or changes to the plan that we are implementing. Let's merge this and iterate on it if we find new stuff that's needs updating
/lgtm
/approve
